The job **`tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/27912747390/job/82592632450) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/82592632450) [(plain)](https://github.com/rust-lang/rust/commit/d3373c9d0d1134bbe95101d21f891d31a61f90ec/checks/82592632450/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[TIMING:end] tool::Tidy { compiler: Compiler { stage: 0, host: x86_64-unknown-linux-gnu, forced_compiler: false }, target: x86_64-unknown-linux-gnu } -- 0.000
fmt check
fmt: checked 6926 files
tidy check
tidy [ui_tests (tests)]: issue-number-only test names are not descriptive, consider renaming file `tests/ui/macros/issue-158216.rs` to `{reason}-issue-158216.rs`
tidy [ui_tests (tests)]: FAIL
tidy [rustdoc_json (src)]: `rustdoc-json-types` modified, checking format version
tidy: Skipping binary file check, read-only filesystem
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
---
info: ✓ ES-Check passed! All files are ES10 compatible.
typechecking javascript files
tidy: The following check failed: ui_tests (tests)
Bootstrap failed while executing `test src/tools/tidy tidyselftest --extra-checks=py,cpp,js,spellcheck`
Command `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools-bin/rust-tidy --root-path=/checkout --cargo-path=/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo --output-dir=/checkout/obj/build --concurrency=4 --npm-path=/node/bin/yarn --ci=true --extra-checks=py,cpp,js,spellcheck` failed with exit code 1
Created at: src/bootstrap/src/core/build_steps/tool.rs:1624:23
Executed at: src/bootstrap/src/core/build_steps/test.rs:1511:29

--- BACKTRACE vvv
   0: <bootstrap::utils::exec::DeferredCommand>::finish_process
             at /checkout/src/bootstrap/src/utils/exec.rs:939:17
   1: <bootstrap::utils::exec::DeferredCommand>::wait_for_output::<&bootstrap::utils::exec::ExecutionContext>
             at /checkout/src/bootstrap/src/utils/exec.rs:831:21
   2: <bootstrap::utils::exec::ExecutionContext>::run
             at /checkout/src/bootstrap/src/utils/exec.rs:741:45
   3: <bootstrap::utils::exec::BootstrapCommand>::run::<&bootstrap::core::builder::Builder>
             at /checkout/src/bootstrap/src/utils/exec.rs:339:27
   4: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1511:29
   5: <bootstrap::core::builder::Builder>::ensure::<bootstrap::core::build_steps::test::Tidy>
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1596:36
   6: <bootstrap::core::build_steps::test::Tidy as bootstrap::core::builder::Step>::make_run
             at /checkout/src/bootstrap/src/core/build_steps/test.rs:1433:21
   7: <bootstrap::core::builder::StepDescription>::maybe_run
             at /checkout/src/bootstrap/src/core/builder/mod.rs:476:13
   8: bootstrap::core::builder::cli_paths::match_paths_to_steps_and_run
             at /checkout/src/bootstrap/src/core/builder/cli_paths.rs:232:18
   9: <bootstrap::core::builder::Builder>::run_step_descriptions
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1139:9
  10: <bootstrap::core::builder::Builder>::execute_cli
             at /checkout/src/bootstrap/src/core/builder/mod.rs:1118:14
  11: <bootstrap::Build>::build
             at /checkout/src/bootstrap/src/lib.rs:803:25
  12: bootstrap::main
             at /checkout/src/bootstrap/src/bin/main.rs:130:11
  13: <fn() as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:250:5
  14: std::sys::backtrace::__rust_begin_short_backtrace::<fn(), ()>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/sys/backtrace.rs:166:18
  15: std::rt::lang_start::<()>::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:206:18
  16: <&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe as core::ops::function::FnOnce<()>>::call_once
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/core/src/ops/function.rs:287:21
  17: std::panicking::catch_unwind::do_call::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
  18: std::panicking::catch_unwind::<i32, &dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:544:19
  19: std::panic::catch_unwind::<&dyn core::ops::function::Fn<(), Output = i32> + core::marker::Sync + core::panic::unwind_safe::RefUnwindSafe, i32>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panic.rs:359:14
  20: std::rt::lang_start_internal::{closure#0}
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/rt.rs:175:24
  21: std::panicking::catch_unwind::do_call::<std::rt::lang_start_internal::{closure#0}, isize>
             at /rustc/0417c25868d6dfbd1c291dfeae950504faa6f790/library/std/src/panicking.rs:581:40
---
  28: __libc_start_main
  29: _start


Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:02:39
  local time: Sun Jun 21 18:03:12 UTC 2026
  network time: Sun, 21 Jun 2026 18:03:12 GMT
##[error]Process completed with exit code 1.
##[group]Run echo "disk usage:"
```

</details>

--- a/compiler/rustc_builtin_macros/src/format.rs
+++ b/compiler/rustc_builtin_macros/src/format.rs
@@ -607,6 +607,8 @@ fn make_format_args(
         // If there's a lot of unused arguments,
         // let's check if this format arguments looks like another syntax (printf / shell).
         let detect_foreign_fmt = unused.len() > args.explicit_args().len() / 2;
+        let foreign_fmt_str =
+            if append_newline { fmt_str.strip_suffix('\n').unwrap_or(fmt_str) } else { fmt_str };
         report_missing_placeholders(
             ecx,
             unused,
@@ -616,7 +618,7 @@ fn make_format_args(
             &invalid_refs,
             detect_foreign_fmt,
             str_style,
-            fmt_str,
+            foreign_fmt_str,
             uncooked_fmt_str.1.as_str(),
             fmt_span,
         );

--- a/tests/ui/macros/format-foreign-dollar-without-spec.stderr
+++ b/tests/ui/macros/format-foreign-dollar-without-spec.stderr
@@ -2,15 +2,13 @@ error: argument never used
   --> $DIR/format-foreign-dollar-without-spec.rs:3:25
    |
 LL |     println!("%65536$", 1);
-   |                         ^ argument never used
+   |              ---------  ^ argument never used
+   |              |
+   |              formatting specifier missing
    |
-note: format specifiers use curly braces, and the conversion specifier `
-      ` is unknown or unsupported
-  --> $DIR/format-foreign-dollar-without-spec.rs:3:15
+help: format specifiers use curly braces, consider adding a format specifier
    |
-LL |     println!("%65536$", 1);
-   |               ^^^^^^^^
-   = note: printf formatting is not supported; see the documentation for `std::fmt`
+LL |     println!("%65536${}", 1);
+   |                      ++
 
 error: aborting due to 1 previous error
-

--- a/tests/ui/macros/format-foreign-dollar-without-spec.stderr
+++ b/tests/ui/macros/format-foreign-dollar-without-spec.stderr
@@ -12,3 +12,4 @@ LL |     println!("%65536${}", 1);
    |                      ++
 
 error: aborting due to 1 previous error
+

--- a/tests/ui/macros/issue-92267.stderr
+++ b/tests/ui/macros/issue-92267.stderr
@@ -1,14 +1,15 @@
 error: argument never used
   --> $DIR/issue-92267.rs:3:34
    |
-LL | pub fn main() { println!("🦀%%%", 0) }
+LL | pub fn main() { println!("🦀%%%", 0) } //~ ERROR argument never used
    |                          -------  ^ argument never used
    |                          |
    |                          formatting specifier missing
    |
 help: format specifiers use curly braces, consider adding a format specifier
    |
-LL | pub fn main() { println!("🦀%%%{}", 0) }
+LL | pub fn main() { println!("🦀%%%{}", 0) } //~ ERROR argument never used
    |                                ++
 
 error: aborting due to 1 previous error
+

--- a/tests/ui/macros/issue-92267.stderr
+++ b/tests/ui/macros/issue-92267.stderr
@@ -1,15 +1,14 @@
 error: argument never used
   --> $DIR/issue-92267.rs:3:34
    |
-LL | pub fn main() { println!("🦀%%%", 0) } //~ ERROR argument never used
+LL | pub fn main() { println!("🦀%%%", 0) }
    |                          -------  ^ argument never used
    |                          |
    |                          formatting specifier missing
    |
 help: format specifiers use curly braces, consider adding a format specifier
    |
-LL | pub fn main() { println!("🦀%%%{}", 0) } //~ ERROR argument never used
+LL | pub fn main() { println!("🦀%%%{}", 0) }
    |                                ++
 
 error: aborting due to 1 previous error
-

--- a/tests/ui/macros/issue-92267.stderr
+++ b/tests/ui/macros/issue-92267.stderr
@@ -2,15 +2,13 @@ error: argument never used
   --> $DIR/issue-92267.rs:3:34
    |
 LL | pub fn main() { println!("🦀%%%", 0) }
-   |                                   ^ argument never used
+   |                          -------  ^ argument never used
+   |                          |
+   |                          formatting specifier missing
    |
-note: format specifiers use curly braces, and the conversion specifier `
-      ` is unknown or unsupported
-  --> $DIR/issue-92267.rs:3:30
+help: format specifiers use curly braces, consider adding a format specifier
    |
-LL | pub fn main() { println!("🦀%%%", 0) }
-   |                               ^^
-   = note: printf formatting is not supported; see the documentation for `std::fmt`
+LL | pub fn main() { println!("🦀%%%{}", 0) }
+   |                                ++
 
 error: aborting due to 1 previous error
-

--- a/tests/ui/macros/trailing-percent-format-hint-issue-158216.rs
+++ b/tests/ui/macros/trailing-percent-format-hint-issue-158216.rs
@@ -1,0 +1,7 @@
+//@ check-fail
+//@ compile-flags: --crate-type=lib
+
+pub fn f(x: f64) {
+    println!("{x:>8.2}%", "foo");
+    //~^ ERROR argument never used
+}

--- a/tests/ui/macros/trailing-percent-format-hint-issue-158216.stderr
+++ b/tests/ui/macros/trailing-percent-format-hint-issue-158216.stderr
@@ -12,3 +12,4 @@ LL |     println!("{x:>8.2}%{}", "foo");
    |                        ++
 
 error: aborting due to 1 previous error
+

--- a/tests/ui/macros/trailing-percent-format-hint-issue-158216.stderr
+++ b/tests/ui/macros/trailing-percent-format-hint-issue-158216.stderr
@@ -1,0 +1,14 @@
+error: argument never used
+  --> $DIR/trailing-percent-format-hint-issue-158216.rs:5:27
+   |
+LL |     println!("{x:>8.2}%", "foo");
+   |              -----------  ^^^^^ argument never used
+   |              |
+   |              formatting specifier missing
+   |
+help: format specifiers use curly braces, consider adding a format specifier
+   |
+LL |     println!("{x:>8.2}%{}", "foo");
+   |                        ++
+
+error: aborting due to 1 previous error

